### PR TITLE
[Proposal] Implement IExternalFileSystem interface.

### PR DESCRIPTION
**Pull request by BenLubar**
This interface does not exist yet, but it would be useful to allow faster downloads of packages by not requiring the use of ProGet as a proxy.

See also: Inedo/inedox-aws#1

### CODE REVIEW

**Comment by apxltd** [comment]
Oh, I had forgotten about this. Actually I had talked to @gdivis about this idea a while back, and I forgot everything he mentioned, but the basic thing was that it would involve a lot more consideration with IExternalFileSystem.

Will all the clients (nuget.exe, npm etc) support "changing" download URLS? How will this work with caching? What about replication?

What's the main usecase for it? How do we configure it? Where does it get listed in the docs, etc.

So basically we'd need to start with a more detailed overview of the feature/specs.

**Comment by BenLubar** [comment]
> Will all the clients (nuget.exe, npm etc) support "changing" download URLS?

NuGet and NPM do, at least. Instructions on how to repro are here: https://gist.github.com/BenLubar/25c235579670a6d89bac36c6002182bb

> How will this work with caching? What about replication?

For files in a package store that implements `IExternalFileSystem`, if an HTTP request would normally read the entire file and then send it to the user verbatim, it can instead redirect to the URL returned from `GetExternalDownloadUrlAsync` to save a round trip.

For caching, on the first request for a specific version of a package, we can send the file to the package store and the client at the same time, assuming nothing needs to access the entire file before it can be sent off.

> What's the main usecase for it? How do we configure it? Where does it get listed in the docs, etc.

`IExternalFileSystem` should be used in exactly one case: ProGet would download a file from an external package store and send it directly to the user with no modifications. Instead, ProGet would send an HTTP 302 response, optionally with a `Cache-Control: max-age=...` header, and the client would load the file directly from Amazon or Azure.

There wouldn't need to be any configuration - extensions that support the interface just automatically work. The two implementations I wrote so far grant temporary download access to the specific file for 6 hours and return a URL with a signature that allows the client to access the file.

The only additional information given to the user over a regular proxied download is the name of the S3 bucket or Azure blob container, which shouldn't be something that needs to be kept secret anyway. Both implementations work properly with secure, locked-down permissions on the cloud data store.

**Comment by gdivis** [comment]
This sounds good to me - one possible use case for us is if we ever decide to start using a ProGet instance for our software/extensions, since we want to redirect downloads to our S3 buckets instead of hosting them all on inedo.com.

Of course docs are always a major consideration, though it seems to be something that could be enabled via a checkbox in the filesystem editor.
